src/hangar/columns/column.py

```diff
@@ -0,0 +1,701 @@
+
+    # ------------------------ Writer-Enabled Methods Only ------------------------------
+
+    @writer_checkout_only
+    def __delitem__(self, key: str) -> str:
+        """remove a column and all data records if write-enabled process.
+
+        Parameters
+        ----------
+        key : str
+            Name of the column to remove from the repository. This will remove
+            all records from the staging area (though the actual data and all
+            records are still accessible) if they were previously committed
+
+        Returns
+        -------
+        str
+            If successful, the name of the removed column.
+
+        Raises
+        ------
+        PermissionError
+            If any enclosed column is opened in a connection manager.
+        """
+        if self._any_is_conman():
+            raise PermissionError(
+                'Not allowed while any columns class is opened in a context manager')
+        return self.delete(key)
+
+    @writer_checkout_only
+    def delete(self, column: str) -> str:
+        """remove the column and all data contained within it.
+
+        Parameters
+        ----------
+        column : str
+            name of the column to remove
+
+        Returns
+        -------
+        str
+            name of the removed column
+
+        Raises
+        ------
+        PermissionError
+            If any enclosed column is opened in a connection manager.
+        KeyError
+            If a column does not exist with the provided name
+        """
+        if self._any_is_conman():
+            raise PermissionError(
+                'Not allowed while any columns class is opened in a context manager')
+
+        with ExitStack() as stack:
+            datatxn = TxnRegister().begin_writer_txn(self._dataenv)
+            stack.callback(TxnRegister().commit_writer_txn, self._dataenv)
+
+            if column not in self._columns:
+                e = KeyError(f'Cannot remove: {column}. Key does not exist.')
+                raise e from None
+
+            column_layout = self._columns[column].column_layout
+            columnSchemaKey = schema_db_key_from_column(column, layout=column_layout)
+            column_record = schema_column_record_from_db_key(columnSchemaKey)
+            startRangeKey = dynamic_layout_data_record_db_start_range_key(column_record)
+
+            self._columns[column]._close()
+            self._columns.__delitem__(column)
+            with datatxn.cursor() as cursor:
+                cursor.first()
+                recordsExist = cursor.set_range(startRangeKey)
+                while recordsExist:
+                    k = cursor.key()
+                    if k.startswith(startRangeKey):
+                        recordsExist = cursor.delete()
+                    else:
+                        recordsExist = False
+            datatxn.delete(columnSchemaKey)
+
+        return column
+
+    @writer_checkout_only
+    def create_str_column(self,
```

Comment:
@rlizzo @lantiga 
IMO; these names are verbose and maybe we could try something else. Few ideas on top of my mind (which needs polishing) are
- `co.columns.create(type='ndarray' / 'str', layout='flat' / 'nested')` where the argument names are arbitrary and could be anything else
- Constructors being nouns like `np.array` -> `co.column.flat`, `co.column.nested`, `col.column.str`

Reply:
I think that verbosity is a feature here; customers need that verbosity when they're trying to learn the API.

I don't think that `co.columns.create` is particularly Pythonic, since the intermediate `columns` object is tied to and operating on the `co`, but presumably is also a dict-like thing. Calling create on a dict to manipulate the dict's parent object? That's weird.  Also, `flat` has no meaning to me, though perhaps that will be fixed once the v0.5.0 docs are live.

Reply:
When I said verbose, I meant, we are repeating information here. The API in this PR is `co.columns.create_xxx_column` and by this, we are repeating the word `column` which would be obvious even if we use `co.columns.create()`.  I am not particularly sentimental about the name `create`. It could be anything else, like `add` or `update` (like `dict.update`, thinking aloud here).

Reply:
Ah, I misunderstood, sorry.

I think the function should go on the checkout object; putting it on the columns pseudo-container seems odd.

Reply:
Hey guys as a follow up for completeness here...

1) The method has been renamed to `add_ndarray_column()` and `add_string_column()`. This is to reflect the actual workings of the method. Me and @lantiga considered `define_foo_column()` as well (as suggested by @elistevens) but decided against it because while acting as a definition, the operation also creates a column ready for use. We wanted to make it clear that nothing else needs to be done in order to use the column that was "defined/initialized" 

2) The method does now live on the `checkout` class (instead of `checkout.columns`). This is just better API design. ex: 

```python
co = repo.checkout(write=True)
col = co.add_ndarray_column('foo', shape=(10, 10), dtype=np.uint8)
col[1] = np.arange(10, dtype=np.uint8)
co.commit('added column')
co.close()
```

3) I'm going to be working on improving docstrings and tutorials shortely, but with the new "generalized" column forms, the options essentially break down into the following:

- `"layout"`: the method by which data in columns are named/indexed/accessed by. This can currently take on the values of `"flat"` (for single-level story type mapping) OR `"nested"` (for a sample name which contains subsample name/data pairs)

- `"type"`: indicates the format of data stored within the column container. Valid values are `"ndarray"` or `"string"`. These are not accepted as keyword arguments, but are determined by the method name which is actually called to create the column (`add_ndarray_column()` for `"ndarray"` and `add_str_column()` for `"string"`)

- `"variable_shape"`: bool indicating if data length / size can be variable sized (will have different defaults and avalability depending on the column `"type"). For example, `ndarray` columns accept both `True` and `False` values, while `string` column types only accept `True`. 

- `"backend"` and `"backend_options"`: specify specific backend and any filter / compression options to apply to the column. These are advanced arguments which are type & value checked / enforced based on the definition of column values described above. I won't get into the full details here... 

Making this all clear is a high priority for me, and i'm going to need input on the explanations we eventually get around to writing. Let me know if you have any questions!